index.d.ts

```diff
@@ -143,14 +143,14 @@ export class Axios {
     response: AxiosInterceptorManager<AxiosResponse>;
   };
   getUri(config?: AxiosRequestConfig): string;
-  request<T = never, R = AxiosResponse<T>> (config: AxiosRequestConfig<T>): Promise<R>;
-  get<T = never, R = AxiosResponse<T>>(url: string, config?: AxiosRequestConfig<T>): Promise<R>;
-  delete<T = never, R = AxiosResponse<T>>(url: string, config?: AxiosRequestConfig<T>): Promise<R>;
-  head<T = never, R = AxiosResponse<T>>(url: string, config?: AxiosRequestConfig<T>): Promise<R>;
-  options<T = never, R = AxiosResponse<T>>(url: string, config?: AxiosRequestConfig<T>): Promise<R>;
-  post<T = never, R = AxiosResponse<T>>(url: string, data?: T, config?: AxiosRequestConfig<T>): Promise<R>;
-  put<T = never, R = AxiosResponse<T>>(url: string, data?: T, config?: AxiosRequestConfig<T>): Promise<R>;
-  patch<T = never, R = AxiosResponse<T>>(url: string, data?: T, config?: AxiosRequestConfig<T>): Promise<R>;
+  request<T = never, R = AxiosResponse<T>> (config: AxiosRequestConfig): Promise<R>;
+  get<T = never, R = AxiosResponse<T>>(url: string, config?: AxiosRequestConfig): Promise<R>;
+  delete<T = never, R = AxiosResponse<T>>(url: string, config?: AxiosRequestConfig): Promise<R>;
+  head<T = never, R = AxiosResponse<T>>(url: string, config?: AxiosRequestConfig): Promise<R>;
+  options<T = never, R = AxiosResponse<T>>(url: string, config?: AxiosRequestConfig): Promise<R>;
+  post<T = never, R = AxiosResponse<T>>(url: string, data?: T, config?: AxiosRequestConfig): Promise<R>;
+  put<T = never, R = AxiosResponse<T>>(url: string, data?: T, config?: AxiosRequestConfig): Promise<R>;
+  patch<T = never, R = AxiosResponse<T>>(url: string, data?: T, config?: AxiosRequestConfig): Promise<R>;
```

Comment:
The idea of allowing `AxiosRequestConfig` to specify the type of `data` was a good one. The real issue is that `T` is used as the type of both the **request body** and the **response body**.

Reply:
That makese sense.

In this case I can see two options here:

1. to omit the type parameter in the `request/get/delete/...` method parameter types,
    ```patch
    -  request<T = never, R = AxiosResponse<T>> (config: AxiosRequestConfig<T>): Promise<R>;
    +  request<T = never, R = AxiosResponse<T>> (config: AxiosRequestConfig): Promise<R>;
    ```
2. to introduce a third type parameter (e.g. `D` for data) there:
    ```patch
    -  request<T = never, R = AxiosResponse<T>> (config: AxiosRequestConfig<T>): Promise<R>;
    +  request<T = never, D = any, R = AxiosResponse<T>> (config: AxiosRequestConfig<D>): Promise<R>;
    ```

What makes more sense in your opinion?

Reply:
I have updated this PR in accordance to the second option.

Having said this, it could make sense to rename the type parameter to `Data` to make it more obvious what it stands for.

Reply:
And based on your input, I have now moved the `D` parameter to the end, to avoid a breaking change.

Reply:
I'd like to voice my support of using a verbose name for the generics in this case, so they're a little more expressive! Hopefully that'll help avoid any issues like #4109 in the future.

Reply:
@remcohaszing opted to rename the generic type parameters separately, and as follows:

> T → ResponseData
> R → ReturnValue
> D → RequestData

If @jasonsaayman prefers this as well, I could make the necessary changes right here in this PR, but since that's a distinct improvement, whereas the current changes are mainly *fix*ing a regression, I would recommend to separate that renaming for better traceabilty.